module/component/component.go

```diff
@@ -169,91 +148,108 @@ func NewComponentManagerBuilder() ComponentManagerBuilder {
-		go func() {
-			select {
-			case err := <-errChan:
-				cancel() // shutdown all workers
-
-				// we propagate the error directly to the parent because a failure in a
-				// worker routine is considered irrecoverable
-				parent.Throw(err)
-			case <-c.done:
-				// Without this additional select, there is a race condition here where the done channel
-				// could be closed right after an irrecoverable error is thrown, so that when the scheduler
-				// yields control back to this goroutine, both channels are available to read from. If this
-				// second case happens to be chosen at random to proceed, then we would return and silently
-				// ignore the error.
-				select {
-				case err := <-errChan:
-					cancel()
-					parent.Throw(err)
-				default:
-				}
-			}
-		}()
+	// Make sure we only start once. atomically check if started is false then set it to true.
+	// If it was not false, panic
+	if !c.started.CAS(false, true) {
+		panic(module.ErrMultipleStartup)
+	}
 
-		var workersReady sync.WaitGroup
-		var workersDone sync.WaitGroup
-		workersReady.Add(len(c.workers))
-		workersDone.Add(len(c.workers))
-
-		// launch workers
-		for _, worker := range c.workers {
-			worker := worker
-			go func() {
-				defer workersDone.Done()
-				var readyOnce sync.Once
-				worker(signalerCtx, func() {
-					readyOnce.Do(func() {
-						workersReady.Done()
-					})
-				})
-			}()
+	ctx, cancel := context.WithCancel(parent)
+	signalerCtx, errChan := irrecoverable.WithSignaler(ctx)
+	go func() {
+		<-ctx.Done()
+		close(c.shutdownSignal)
+	}()
+
+	// launch goroutine to propagate irrecoverable error
+	go func() {
+		// wait until the done channel is closed or an irrecoverable error is encountered
+		doneCtx, _ := util.WithDone(context.Background(), c.done)
+		if err := util.WaitError(doneCtx, errChan); err != nil {
+			cancel() // shutdown all workers
+
+			// propagate the error directly to the parent because a failure in a worker routine
+			// is considered irrecoverable
+			parent.Throw(err)
 		}
+	}()
 
-		// launch goroutine to close ready channel
-		go c.waitForReady(&workersReady)
+	var workersReady sync.WaitGroup
+	var workersDone sync.WaitGroup
+	workersReady.Add(len(c.workers))
+	workersDone.Add(len(c.workers))
 
-		// launch goroutine to close done channel
-		go c.waitForDone(&workersDone)
-	} else {
-		panic(module.ErrMultipleStartup)
+	// launch workers
+	for _, worker := range c.workers {
+		worker := worker
+		go func() {
+			var readyOnce sync.Once
+			worker(signalerCtx, func() {
+				readyOnce.Do(func() {
+					workersReady.Done()
+				})
+			})
+			// This must be called last. It can't be called in a defer statement because thrown
+			// irrecoverable errors cause an os.Goexit to be called, which will call all defers.
+			// This creates a race condition where this component manager could be marked done
+			// before the parent reads the error off the channel resulting in unsafe continuation.
+			workersDone.Done()
```

Comment:
This isn't quite accurate.

If an irrecoverable error is thrown, the error channel is guaranteed to close *before* the done channel closes. It is up to the parent to always check if there is an error, even if the done channel is closed. This is the reason why we defined the `WaitError` function.

Reply:
This is the main requested change, most other comments are minor.

Reply:
I *think* it's because we propagate the error from `ComponentManager` to the parent in a separate goroutine. So if the parent gets scheduled before the error handler goroutine, the parent could see the done signal before the error.

when running the node unit tests with `-count=500`, it always failed during `Run encounters irrecoverable error during shutdown` because the logs showed it shutdown gracefully instead of with an error. Moving that `Done()` out of the defer fixed it.

I can update the comment to make that more clear.

Reply:
this is another class of race conditions we'll have to consider, since I suspect we'll have multiple layers of chaining in some cases

Reply:
So, the reason that moving the `Done` out of the defer fixed that specific test is because the `Done` just never gets called since `os.GoExit` is called. However, this is not the right way to address this issue, since we always want `Done` to be closed.

The correct way to fix this is to add the `WaitError` (or similar logic) to the test instead.

Reply:
Also, can you run the `component_manager_test` manual test with the changes here?